### DISCUSSION TITLE
Even if pulse is coming in the PM5 why?

### DISCUSSION THREAD

**Discussion by JDALIZSTUDIO** [JaapvanEkris/openrowingmonitor#47] [General]
I did a recoding session of the Concept2 D and replay it using a Cypress Device the signal is always 5V and The math was exaclly the same on the PM5. But I noticed that when the Pulse frequency is not reach a certain frequency the distance is not increasing. I don't see that in the Math where is it coming from?

**Comment by JaapvanEkris** [comment]
It is decided by Rower.js, which requires that flywheel.isAboveMinimumSpeed() before it will leave the state "waitingForDrive". Every monitor does that.

**Reply by JDALIZSTUDIO** [reply]
Thanks. But sounds like it use Power for this not just speed. I observed that the frequency when it start is for example 20Hz (The PM5 start increasing distance) then i reduce the speed to towards 25Hz it stop. I have a Signal Generator that can simulate the Concept2 signal.

**Reply by JaapvanEkris** [reply]
> Thanks. But sounds like it use Power for this not just speed. I observed that the frequency when it start is for example 20Hz (The PM5 start increasing distance) then i reduce the speed to towards 25Hz it stop.

No, because Power also includes Angular Acceleration ([see the Physics Model explanation for more info](https://github.com/JaapvanEkris/openrowingmonitor/blob/main/docs/physics_openrowingmonitor.md)).

It is interesting that at 20Hz the PM starts (i.e. the config setting maximumTimeBetweenImpulses: 0.0145 in OpenRowingMonitor). Please note that ORM matches the PM5 structurally within 0.1% (with maximumTimeBetweenImpulses: 0.02 and a modified regression algorithm), so the app and config are pretty good at matching the PM5's behaviour on a real rowing machine. As we are still optimizing the config to the new algorithms, this number might change (as we aim to bring the deviation down to 0.01%). So the 0.04 sec maximumTimeBetweenImpulses might be a realistic number which I'll simulate later on in the optimisation process.

It actually is surprising that it stops at 25Hz, as the higher frequency implies that the simulated flywheel is actually spinning faster. It could be that it times out: after 6 seconds without a drive, a PM5 will stop. A PM5 will look for a very specific pattern, as it is a frequency modulated sinoid when flywheel speed is constant, and it should contain an acceleration (i.e. the drive) and a deceleration (i.e. the recovery) at least every 6 seconds. So the PM5 might just have detected that the signal isn't realistic.

**Comment by ShamaiInsight** [comment]
[DSLogic-la-200721-170535.txt](https://github.com/user-attachments/files/15838491/DSLogic-la-200721-170535.txt)
[DSLogic-la-200721-170535.csv.Speed.csv](https://github.com/user-attachments/files/15838493/DSLogic-la-200721-170535.csv.Speed.csv)
Here are the recorded Data the speed.csv from (DSLogic is a series of USB-based logic analyzer, with max sample rate up to 1GHz, and max sample depth up…) i convert it using Matlab script to generate a list of frequency you can do a 1/Frq to have the pulse period
The .txt is the result on the PM5

**Comment by ShamaiInsight** [comment]
![image](https://github.com/JaapvanEkris/openrowingmonitor/assets/124437297/1e02d157-7534-42af-bd36-d4d2b9707ae2)
Before implementing on my CPU which have an Input Capture it's a PIC24 i wanted to check if i can get an algorithm working
I have not fully finished I need to add the Cal the Split and Energy
[Updated_C2FullAlgo.zip](https://github.com/user-attachments/files/15838637/Updated_C2FullAlgo.zip)

I have dificulty with the Math I must admit

**Reply by ShamaiInsight** [reply]
I got this result when using the 170535.csv.Speed.csv capture
 
![image](https://github.com/JaapvanEkris/openrowingmonitor/assets/124437297/3b7e7b95-7393-4398-a63a-6a0009f57658)

![image](https://github.com/JaapvanEkris/openrowingmonitor/assets/124437297/2fe0de4e-71bb-4e63-b7d5-650dd6216c4a)

**Reply by JaapvanEkris** [reply]
> I have dificulty with the Math I must admit

What are you trying to achieve? Because the math to transform a raw signal into decent metrics is pretty hefty and for a Concept2 we are pretty close to maxing out the CPU. The math, physics and computer science involded is far from trivial. To get this working on any platform, some thorough understanding of the math and physics involed is quite essential.

This code isn't easy to port to be honest, given the high demands on the CPU. Please realize you need nanosecond accurate measurements and you need to process them realtime for metrics to be usefull. I'm still amazed we managed to do it using a PREEMPT kernel and Javascript, but our sister project uses C++ for very good reasons. And that sister project is run by someone who also has a huge experience in embedded systems development. So porting this is a far from trivial task (and installing this project on a Pi Zero 2W or our sister project on a ESP32 is a lot easier).

**Reply by JaapvanEkris** [reply]
> I got this result when using the 170535.csv.Speed.csv capture

Are you using a smoothing algorithm and then use velocity = dx / dt? Because on the modern C2's that isn't stable enough to get proper second derivative (i.e angular acceleration) which is required for reliable and responsive Torque calculations. ORM uses a running quadratic Theil-Sen estimator in combination with Weighed averages to achieve that.

**Reply by JDALIZSTUDIO** [reply]
I need to port the Python Code into C code because I need to do a PM5 compatible device which return trough USB Math Stroke Rate, distance, Total Energy, drag Factor. Is their an implementation of the quadratic Theil-Sen estimator in Python or C?

**Reply by JDALIZSTUDIO** [reply]
do I need to Torque to compute the distance and Calories?

**Reply by JaapvanEkris** [reply]
> I need to port the Python Code into C code because I need to do a PM5 compatible device which return trough USB Math Stroke Rate, distance, Total Energy, drag Factor.

We use JavaScript.

OpenRowingMonitor is nearly identical to a PM5 (as said +/- 0.1%, based on 3000 Km of rowing, and we can simulate a PM5 BLE interface quite well). Adding a USB output to the current code is extremely simple, why port it?

> Is their an implementation of the quadratic Theil-Sen estimator in Python or C?

There is. Our sister project has ported that algorithm to C++ to run on an ESP32 (see https://github.com/Abasz/ESPRowingMonitor). And @Abasz maintains the code well and contributed directly and indirectly to this codebase as well. Why not use his ESP32 solution?

**Reply by JaapvanEkris** [reply]
> do I need to Torque to compute the distance and Calories?

You need Torque for reliable stroke detection, as torquePresent() is an important ingredient in it. In our development and calibration tests we see a change in stroke detection accuracy that affects drag calculations when the parameter is set wrong with the current configuration parameters. These errors in drag calculation result in distances being too long or too short with respect to a PM5.

**Reply by Abasz** [reply]
> I need to port the Python Code into C code because I need to do a PM5 compatible device which return trough USB Math Stroke Rate, distance, Total Energy, drag Factor. Is their an implementation of the quadratic Theil-Sen estimator in Python or C?

I dont fully understand what you are trying to do. What is the end goal. PM5 has an SDK and it is able to communicate via USB:
https://www.concept2.com/service/software/software-development

Recently they introduced RasPiRowing which allows connection to Rpi. So really if you want the same data that PM5 already produces, I see little point in reinventing the wheel.

I would like to also add that ESPRowingMonitor's algorithm can be compiled against windows with GCC (make e2e) that allows simulating sessions on PC, I think it can be compiled against RPi too, but probably will need some changes to have all interrupts and everything as it uses Arduino framework and espriff. Technically if the arduino/esp layer is changed to Rpi, I am pretty sure it can work on Rpi too using its GPIO.

EDIT:

I am working on a massive update that will deprecate the wifi and web socket based connection layer and use bluetooth only for everything (keeping the features). It also fixes some bugs, of which one impacts weighted average calculations (that impacts the algorithm). If you want to use it, let me know and I publish the bug fixes as a hot fix release.

**Reply by JaapvanEkris** [reply]
> It also fixes some bugs, of which one impacts weighted average calculations (that impacts the algorithm). If you want to use it, let me know and I publish the bug fixes as a hot fix release.

As an aside: in the Regression_completeness I added an update to the TS estimator that further reduces CPU load (see https://github.com/JaapvanEkris/openrowingmonitor/tree/v1beta_updates_regression_completeness/app/engine/utils). In essence, I introduced lazy calculation of the secondary parameters. As the Quadratic variant pushes a lot of datapoints on the linear version to resolve the linear residu, the old version would also calculate the linear b in the equation for every datapoint added. As many of these parameters are only used sporadically, it removes a lot of unnecessary calculations.

The regression completeness version will eventually update main (a.k.a. V1Beta), as soon as I am able to identify the optimal configuration. Aside from fixing memory leaks (yeah, JavaScript is worse than C++) and performance there are subtle differences in stroke detection as I'm moving towards using the TS estimatir for stroke detection as well. Due to the many simulations, this will take several months.

**Reply by JDALIZSTUDIO** [reply]
I have another constraint: the board is already built and has a USB stack already implement with a specific protocol, so I can't make it communicate with the PM5. It needs to be a replacement for the PM5. However, I didn't realize that you have a C++ version. That might help me a lot. Thanks

**Reply by JaapvanEkris** [reply]
> the board is already built and has a USB stack already implement with a specific protocol, so I can't make it communicate with the PM5. It needs to be a replacement for the PM5.

Again, why do you want to replace a PM5? Why replace a device that is the golden standard? Getting the data out via a websocket is an easy job on OpenRowingMonitor and ESPRowingMonitor. If you are looking for gathering more data, OpenRowingMonitor is structurally validated against a PM5 and can provide extremely detailed metrics after a session. If you are looking for a cheaper alternative to replace a broken PM5, an ESP32 is about €19, and a Pi Zero 2W with a header is about the same. So both solutions would work. Please note, on a RowErg, you need additional electronics as the output voltage is 18Volt.

**Reply by Abasz** [reply]
> As an aside: in the Regression_completeness I added an update to the TS estimator that further reduces CPU load (see https://github.com/JaapvanEkris/openrowingmonitor/tree/v1beta_updates_regression_completeness/app/engine/utils). In essence, I introduced lazy calculation of the secondary parameters. As the Quadratic variant pushes a lot of datapoints on the linear version to resolve the linear residu, the old version would also calculate the linear b in the equation for every datapoint added. As many of these parameters are only used sporadically, it removes a lot of unnecessary calculations.

I've been following these changes on the calculation, I have originally implemented in a somewhat "lazy" way. To be precise before moving to the weighted average I did not even calculate many of the values (e.g. A and B for the TS linear) as they were not used at all (there was no goodness of fit for the quadratic TS). Now that the weighted average needs goodness of fit from the quadratic, so slope and other things are needed I added flags for invalidation (I dont like invalidation via 'null' not to mention that does not really work in statically typed language C++). Also, since TS Quadratic uses goodness of fit for every new data point meaning that it needs all slope, A, B, C on every data point I did not add "lazy" evaluation (though everything is calculated once on adding a new data point) to that as it by definition cannot delay calculation.

> The regression completeness version will eventually update main (a.k.a. V1Beta), as soon as I am able to identify the optimal configuration. Aside from fixing memory leaks (yeah, JavaScript is worse than C++) and performance there are subtle differences in stroke detection as I'm moving towards using the TS estimatir for stroke detection as well. Due to the many simulations, this will take several months.

Why do you need to change the setting? as far as I understand the regression model did not change right?

**Reply by JaapvanEkris** [reply]
> I've been following these changes on the calculation, I have originally implemented in a somewhat "lazy" way. To be precise before moving to the weighted average I did not even calculate many of the values (e.g. A and B for the TS linear) as they were not used at all (there was no goodness of fit for the quadratic TS).

Smart move, you were often a step ahead of me already :)

> Now that the weighted average needs goodness of fit from the quadratic, so slope and other things are needed I added flags for invalidation (I dont like invalidation via 'null' not to mention that does not really work in statically typed language C++).

It isn't production code yet. Although in JavaScript I can get away with it and it saves a lot of weird variables. But I'm looking at more elegant solutions.

> Also, since TS Quadratic uses goodness of fit for every new data point meaning that it needs all slope, A, B, C on every data point I did not add "lazy" evaluation (though everything is calculated once on adding a new data point) to that as it by definition cannot delay calculation.

That is true in the current implementation. I'm experimenting the use of the Quadratic TS for the stroke detection, and then you don't need a C and GoodnessOfFit on every datapoint (actually, when you are at the begin of a drive or recovery or the Torque is wrong, you don't even need an A and B). It is complex and CPU intensive, but it might yield results structurally closer to a PM5.

> Why do you need to change the setting? as far as I understand the regression model did not change right?

Stroke detection currently is (partially) based on OLS for identifying the slope of the currentDt's through time, essentially looking for a slope close to the slope seen in a recovery. Disadvantage is that the recovery slope is determined using LinearTS, which is much more robust to noise, potentially breaking stroke detection due to too much noise. By using LinearTS for stroke detection, I restore the symmetry, but it also brings subtle changes to the moment when a drive and recovery are detected. As stroke detection becomes less vulnerable to data noise, it behaves more reliable. It also allows for different configurations to work, which used to break on the OLS-based setup due to distractions introduced by noise.

Thing is that small changes in stroke detection also impact drag calculation. As we are pretty close to the PM5-data already (structurally within +/- 0.1%), improving on that requires these small improvements and subtle parameter changes to get closer. As an illustration, the current main-branch has a 0.05% standard deviation with respect to a PM5. The current regression_completeness-branch with a slightly different configuration has a StDev of 0.0335%.

**Reply by ShamaiInsight** [reply]
The ESP32 is ok. I'm trying to figure out how to compute the calories now. I can see only Power in the ESP code.
Also on the Concept2 the formula for Power is based on the Pace. I don't understand what is used on the Pm5 is it watts = 2.8/(split/500)³
or
P=k*ω^3

**Reply by Abasz** [reply]
EspRowingMonitor does not calculate calories. It is explained in the readme. Please have a look at the fork if wyattwww. He added calories

Power is not based on pace, it's the other way around. The formula in the concept2 website is to calculate an average power which is doable because pace is based in power and a magic constant (that is the conversion/correlation between power and speed )

**Reply by Abasz** [reply]
> Stroke detection currently is (partially) based on OLS for identifying the slope of the currentDt's through time, essentially looking for a slope close to the slope seen in a recovery. Disadvantage is that the recovery slope is determined using LinearTS, which is much more robust to noise, potentially breaking stroke detection due to too much noise. By using LinearTS for stroke detection, I restore the symmetry, but it also brings subtle changes to the moment when a drive and recovery are detected. As stroke detection becomes less vulnerable to data noise, it behaves more reliable. It also allows for different configurations to work, which used to break on the OLS-based setup due to distractions introduced by noise.
> 
> Thing is that small changes in stroke detection also impact drag calculation. As we are pretty close to the PM5-data already (structurally within +/- 0.1%), improving on that requires these small improvements and subtle parameter changes to get closer. As an illustration, the current main-branch has a 0.05% standard deviation with respect to a PM5. The current regression_completeness-branch with a slightly different configuration has a StDev of 0.0335%.

So the reason behind my question is that I saw that you use TS linear for drag factor, but OLS is used with the _deltaTimes (which also provides a slope of the _deltaTimes) and I thought in the above you were referring to that.

For ESPRowingMonitor recovery drag factor cannot be calculated with TS Linear, simply because on a long recovery it could eat all the heap (e.g. with 200 data points the full TS Linear would need 128kb - assuming double precision - out of the total 320, and there are other things like BLE and such that would also need memory). So it would create an unstable application. Of course there are ways around (e.g. limit the recovery drag factor period to something that would cap the number of datapoints), but I decided it is not worth it.

Now on the _deltaTimes, I believe that that could be something that is changable, as the series length there is capped by the flankLength. Have you experimented with that ? I.e. using TS Linear for calculating the general _deltaTimes slope instead of OLS?

I also liked the idea of using goodness of fit for weight on the drag factor (especially since it does not create additional material overhead, the goodness of fit is calculated any way) as for instance on the kayak first ergometers (the one that I use are very old, not well maintained and there is a lot of bearing friction creating deviation that is dependent on the speed...) drag goodness of fit had to be reduced to approx. 0.7 (which is very low...) and the drag factor jumps around quite a lot. But I have not had the time to start some testing (especially the effect on the responsiveness of the model to resistance change - on the C2 this is picked up within 2, max 3 strokes and the longer the applied smoothing the slower the responsiveness).

**Reply by JaapvanEkris** [reply]
> So the reason behind my question is that I saw that you use TS linear for drag factor, but OLS is used with the _deltaTimes (which also provides a slope of the _deltaTimes) and I thought in the above you were referring to that.

I am. In theory, the TSLinear for dragfactor (and thus expected recovery slope) would ignore noise that the OLS _deltaTimes can't (partially due to the much shorter flank, partially due to the more noise sensitive algorithm), leading to missed strokes. It currently isn't a practical concern, but it does limit the configuration as some parameters are known to cause this. And based on some experiments I also see that these settings might result in a closer match with a PM5.
 
> For ESPRowingMonitor recovery drag factor cannot be calculated with TS Linear, simply because on a long recovery it could eat all the heap (e.g. with 200 data points the full TS Linear would need 128kb - assuming double precision - out of the total 320, and there are other things like BLE and such that would also need memory). So it would create an unstable application. Of course there are ways around (e.g. limit the recovery drag factor period to something that would cap the number of datapoints), but I decided it is not worth it.

Aside that, it is quite a hefty calculation, as you need to determine the slope between the 200th datapoint and all 199 preceding ones. On a RPi it is doable, but CPU-load and memory constraints are an always present concern.

I agree that the difference between OLS and TS for recovery slope is marginal. Only in edge cases, where parts of the drive get included into the recovery (adding some acceleration into the mix where the drag calculation assumes there only is a deceleration due to drag) you get some instability. Due to the number of datapoints in a recovery, this tends to quite limited, but experiments show that to move beyond the 0.0335% standard deviation, we need some additional tricks.

> Now on the _deltaTimes, I believe that that could be something that is changable, as the series length there is capped by the flankLength. Have you experimented with that ? I.e. using TS Linear for calculating the general _deltaTimes slope instead of OLS?

That is the current experiment I'm running. It works well. It brings the stDev with a PM5 from 0.05% to 0.0335%, so it will be a keeper. But this could also be caused by restoring the symmetry between the two algorithms. CPU-load isn't to insane, especially when you organize the requests in a way that lazy evaluation of if-statements pushes the CPU-intensive stuff to the end.
